tests/qnn/test_iqp_expval.py

```diff
@@ -0,0 +1,177 @@
+            "local_gates",
+            [0.3, 0.2],
+            2,
+            False,
+            False,
+            10_000,
+            10_000,
+            10_000,
+            False,
+        ),
+        (
+            [1],
+            "local_gates",
+            [0.3],
+            1,
+            False,
+            True,
+            10_000,
+            10_000,
+            10_000,
+            False,
+        ),
+        (
+            [[0, 1], [0, 1]],
+            "local_gates",
+            [0.3, 0.2],
+            2,
+            False,
+            True,
+            10_000,
+            10_000,
+            10_000,
+            False,
+        ),
+        ([[1, 0], [0, 1]], "multi_gens", [-0.41], 2, True, False, 10_000, None, None, True),
+        ([[1, 0], [1, 0]], "multi_gens", [0.2], 2, True, True, 10_000, None, None, True),
+    ],
+)
+def test_expval(
+    ops,
+    gates_fn,
+    params,
+    n_qubits,
+    spin_sym,
+    sparse,
+    n_samples,
+    max_batch_samples,
+    max_batch_ops,
+    indep_estimates,
+):  # pylint: disable=too-many-arguments
+    import jax
+    import jax.numpy as jnp
+
+    gates = local_gates(n_qubits, 1)
+    if gates_fn == "multi_gens":
+        gates = [[gates[0][0], gates[1][0]]] + gates[2:]
+
+    key = jax.random.PRNGKey(np.random.randint(0, 99999))
+
+    if not isinstance(ops, csr_matrix):
+        ops = jnp.array(ops)
+
+    exp_val, std = iqp_expval(
+        ops=ops,
+        n_samples=n_samples,
+        key=key,
+        num_wires=n_qubits,
+        pattern=gates,
+        weights=params,
+        spin_sym=spin_sym,
+        sparse=sparse,
+        indep_estimates=indep_estimates,
+        max_batch_samples=max_batch_samples,
+        max_batch_ops=max_batch_ops,
+    )
+
+    dev = device("default.qubit")
+
+    @qnode(dev)
+    def iqp_circuit(weights, pattern, spin_sym, n_qubits, ops):
+        IQP(weights, n_qubits, pattern, spin_sym)
+
+        expectation_operators = []
+        if not isinstance(ops, csr_matrix):
+            for l in ops:
+                for i, qubit in enumerate(l):
+                    if qubit == 1:
+                        expectation_operators.append(expval(PauliZ(i)))
+        else:
+            rows, cols = ops.nonzero()
+            for row, col in zip(rows, cols):
+                if ops[row, col] == 1:
+                    expectation_operators.append(expval(PauliZ(col)))
+
+        return expectation_operators
+
+    if len(ops.shape) == 1:
+        ops = ops.reshape(1, -1)
+
+    simulated_exp_val = jnp.array(iqp_circuit(params, gates, spin_sym, n_qubits, ops))
+
+    for i, val in enumerate(simulated_exp_val):
+        # Due to the distribution, we expect the simulated and the approximated values to be within 2 standard
+        # deviations 96% of the time. We can instead check they are withing 3 standard deviations, which should
+        # be True 99.8% of the time, minimizing stochastic failures.
```

Comment:
So do we want to occasionally see it to fail (e.g. ftqc tests, which have quite some to expect to fail within CI) or not (then we would prefer to seed it globally)